department-of-veterans-affairs/caseflow
# [Reader] Implement error state for annotations

Connect #1854.

I created more unit tests for this because it would have been a pain to test backend failure cases in rspec, and because I suspect it would have been challenging to do drag and drop.

It's not great UI to have only a single error message for all the different actions that could fail. For example, if you edit two comments, and one fails and the other succeeds, you'll see a single error message. Which is it for? However, @abbyraskin said that we should move ahead with this UX for now, since it's an improvement over what we have currently, and users are currently impacted by not having visibility into errors here.

Gif showing some of the functionality:

![annotation errors](https://cloud.githubusercontent.com/assets/829827/25965767/f036f14e-3656-11e7-8e4b-334fad8f015e.gif)

## To Test
1. For each verb in "create", "edit", "delete", "dragging and dropping annotation icon to move it":
    1. See that when you take an action in the UI, the UI is updated immediately, without waiting for the server
    1. See that if the backend succeeds, nothing happens. The UI still correctly reflects the backend state.
    1. See that if the backend fails, an error message appears.
        1. When you click "Back to all documents", and then go back to the document you were viewing, the error message should be gone.
         1. If you make another request and it succeeds, the error message should be gone.
